README.md
# Time Since README

Use Time Since to see how long has passed since you last did a task.

## Features

Set the name of the activity you are timing. Time Since will create a simple timer on the status bar to show you how long it has been since you last did that activity. The text has 3 colors: green, yellow, and red, to show you at a quick glance how long it has been since you did your activity. Green is the 'safe' zone, yellow is the 'average' zone, and red is the 'bad' zone. Try to complete your activity again before the red zone!

Time Since defaults to using 3 preselected colors, however you can change the colors to whatever you want, use your integrated terminals ansiGreen, ansiYellow, and ansiRed colors, or remove the colorizing option completely.

### Green Text Color

![green-img](https://github.com/williamv1217/time-since/blob/master/images/green.png?raw=true)

### Yellow Text Color

![yellow-img](https://github.com/williamv1217/time-since/blob/master/images/yellow.png?raw=true)

### Red Text Color

![red-img](https://github.com/williamv1217/time-since/blob/master/images/red.png?raw=true)

### No Text Color (uses your default status bar color)

![nocolor-img](https://github.com/williamv1217/time-since/blob/master/images/no-color.png?raw=true)

## Requirements

VS Code

## Extension Settings

This extension contributes the following settings:

- `timesince.activityName`: name of your task
- `timesince.useTerminalColors`: use default terminal colors for text output
- `timesince.noColor`: Time Since will not colorize your text based on the time, the text color will be the same as the statusbar text color
- `timesince.textColor.green`: set color for green text
- `timesince.textColor.yellow`: set color for yellow text
- `timesince.textColor.red`: set color for red text
- `timesince.limitTime.average`: set time (in minutes) where the 'average' range starts
- `timesince.limitTime.bad`: set time (in minutes) where the 'bad' range starts

## Known Issues

## Release Notes

### 1.0.1

Updated display name and readme

### 1.0.0

Initial release

## Acknowledgements

Icon by Freepik from [www.flaticon.com](https://www.flaticon.com/)

---


CHANGELOG.md
# Change Log

All notable changes to the "timesince" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release


src/Timer.ts
import * as vscode from "vscode";
import Utils from "./utils";

class Timer {
  running: boolean;
  timer: any;
  statusBarItem: vscode.StatusBarItem;

  constructor() {
    this.running = false;
    this.statusBarItem = vscode.window.createStatusBarItem(
      vscode.StatusBarAlignment.Left
    );
    this.statusBarItem.command = "timesince.resetTimer";
  }

  utils = new Utils();

  startTimer() {
    if (this.running) {
      console.log("Already Running!");
      return;
    }
    console.log("starting the timer");

    this.statusBarItem.color = this.utils.getColor(0);//this.utils.getColor(0);
    this.statusBarItem.text = "Starting TimeSince Timer...";
    const dateTime = new Date();
    const startingTime = dateTime.toString();

    this.running = true;
    this.statusBarItem.show();

    this.timer = setInterval(() => {
      let hours = this.utils.getTime(startingTime).hours;
      let minutes = this.utils.getTime(startingTime).minutes;
      let seconds = this.utils.getTime(startingTime).seconds;
      let hoursToMins = hours * 60;

      this.statusBarItem.color = this.utils.getColor(hoursToMins + minutes);
      this.statusBarItem.text = this.utils.getText(hours, minutes, seconds);
    }, 1000);
  }

  stopTimer() {
    if (!this.running) {
      return;
    }
    console.log("stopping the timer");
    clearInterval(this.timer);
    this.running = false;
    this.statusBarItem.hide();
  }

  resetTimer() {
    if (!this.running) {
      return;
    }
    console.log("resetting the timer");
    this.stopTimer();
    this.startTimer();
  }
}

export default Timer;


src/extension.ts
import * as vscode from "vscode";
import Timer from "./Timer";

const timesinceconfig = vscode.workspace.getConfiguration("timesince");

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
  // Use the console to output diagnostic information (console.log) and errors (console.error)
  // This line of code will only be executed once when your extension is activated
  console.log('Congratulations, your extension "timesince" is now active!');

  const timer = new Timer();
  let startTimer = vscode.commands.registerCommand(
    "timesince.startTimer",
    () => {
      timer.startTimer();
    }
  );

  let stopTimer = vscode.commands.registerCommand("timesince.stopTimer", () => {
    timer.stopTimer();
  });

  let resetTimer = vscode.commands.registerCommand(
    "timesince.resetTimer",
    () => {
      vscode.window
        .showQuickPick(["Yes", "No"], {
          placeHolder: "Do you really want to reset the TimeSince timer?",
          canPickMany: false,
        })
        .then((pick) => {
          if (pick === "Yes") {
            timer.resetTimer();
            vscode.window.showInformationMessage(
              "TimeSince timer has been reset!"
            );
          }
        });
    }
  );

  context.subscriptions.push(startTimer);
  context.subscriptions.push(stopTimer);
  context.subscriptions.push(resetTimer);
}

// this method is called when your extension is deactivated
export function deactivate() {}


src/utils.ts
import * as vscode from "vscode";

class Utils {
  getTime(startingTime: string) {
    let currentTime = new Date().toString();
    let totalTime = Date.parse(currentTime) - Date.parse(startingTime);
    let hours = Math.floor((totalTime / 1000 / 3600) % 60);
    let minutes = Math.floor((totalTime / 1000 / 60) % 60);
    let seconds = Math.floor((totalTime / 1000) % 60);

    return { hours, minutes, seconds };
  }

  getText(hours: number, minutes: number, seconds: number) {
    let config = vscode.workspace.getConfiguration("timesince");
    let baseText = `Last ${config.activityName}:  `;

    var hourText = "";
    var minuteText = "";
    var secondText = "";

    switch (true) {
      case hours === 1:
        hourText = "hour";
        break;
      case hours > 1:
        hourText = "hrs";
        break;
      default:
        hourText = "";
    }

    switch (true) {
      case minutes === 1:
        minuteText = "minute";
        break;
      case minutes > 1:
        minuteText = "mins";
        break;
      default:
        minuteText = "";
    }

    switch (true) {
      case seconds === 1:
        secondText = "second";
        break;
      case seconds > 1:
        secondText = "seconds";
        break;
      default:
        secondText = "";
    }

    var changedText = "";

    if (hours >= 1 && minutes >= 1) {
      changedText = `${hours} ${hourText} ${minutes} ${minuteText} ago`;
    } else if (hours >= 1) {
      changedText = `${hours} ${hourText} ago`;
    } else if (minutes >= 1) {
      changedText = `${minutes} ${minuteText} ago`;
    } else {
      changedText = `${seconds} ${secondText} ago`;
    }
    return baseText + changedText;
  }

  getColor(minutes: number) {
    var green;
    var red;
    var yellow;
    let config = vscode.workspace.getConfiguration("timesince");
    let statusBarForeground = new vscode.ThemeColor("statusBar.foreground");

    if (config.noColor) {
      return statusBarForeground;
    } else if (config.useTerminalColors) {
      green = new vscode.ThemeColor("terminal.ansiGreen");
      red = new vscode.ThemeColor("terminal.ansiRed");
      yellow = new vscode.ThemeColor("terminal.ansiYellow");
    } else {
      green = config.textColor.green;
      red = config.textColor.red;
      yellow = config.textColor.yellow;
    }

    if (minutes < config.limitTime.average) {
      return green;
    } else if (minutes < config.limitTime.bad) {
      return yellow;
    } else {
      return red;
    }
  }
}

export default Utils;
